### PR TITLE
Swss.common.multi.consumer

### PR DESCRIPTION
**Pull request by hrachya-mellanox**

### CODE REVIEW

**Comment by stcheng** [comment]
@lguohan @eladraz

File: common/consumertable.cpp [hunk 1]
```diff
@@ -37,7 +37,10 @@ ConsumerTable::ConsumerTable(DBConnector *db, string tableName) :
         }
         catch (...)
         {
-            delete m_subscribe;
+            if(m_subscribe) {
```
Comment (kcudnik) [inline common/consumertable.cpp]: on constructor entry you could add m_subscribe = NULL;

File: common/select.cpp [hunk 2]
```diff
@@ -8,17 +10,28 @@ using namespace std;
 
 namespace swss {
 
-void Select::addSelectable(Selectable *c)
+void Select::addSelectable(_in_ Selectable *selectable)
 {
-    m_objects.push_back(c);
+    if(std::find(m_objects.begin(), m_objects.end(), selectable) != m_objects.end()) {
+        SWSS_LOG_WARN("Selectable:%p already been added to the list, ignoring\n.");
```
Comment (kcudnik) [inline common/select.cpp]: "\n." is not needed, we can enforce in logs to add \n at the end but currently it goes to syslog so no need

File: common/table.h [hunk 3]
```diff
@@ -38,11 +38,11 @@ class Table {
     void delField(std::string key, std::string field);
 
     virtual ~Table();
+    std::string getTableName()const {return m_tableName;};
```
Comment (kcudnik) [inline common/table.h]: lets keep all code in cpp files ?

Comment (eladraz) [inline common/table.h]: This function isn't in used. Why do you need it?

**Comment by eladraz** [comment]
In general, you should break this commit to 3 differet commits. There are too many logical changes here.

File: common/consumertable.cpp [hunk 4]
```diff
@@ -37,7 +37,10 @@ ConsumerTable::ConsumerTable(DBConnector *db, string tableName) :
...
+                delete m_subscribe;
+                m_subscribe = NULL;
```
Comment (eladraz) [inline common/consumertable.cpp]: There is no needed to "if (m_subscribe)", it's ok to delete NULL.
As for "m_subscribe = NULL;" it's a good change.

File: common/consumertable.cpp [hunk 5]
```diff
@@ -47,7 +50,9 @@ ConsumerTable::ConsumerTable(DBConnector *db, string tableName) :
 
 ConsumerTable::~ConsumerTable()
 {
-    delete m_subscribe;
+    if(m_subscribe) {
+        delete m_subscribe;
+    }
```
Comment (eladraz) [inline common/consumertable.cpp]: This change is not needed. It's ok to delete NULL in C++

File: common/scheme.h [hunk 6]
```diff
@@ -1,17 +1,32 @@
...
+#define APP_INTF_TABLE_NAME             "INTF_TABLE"
+#define APP_NEIGH_TABLE_NAME            "NEIGH_TABLE"
+#define APP_ROUTE_TABLE_NAME            "ROUTE_TABLE"
+
+#define APP_TC_TO_QUEUE_MAP_TABLE_NAME  "TC_TO_QUEUE_MAP_TABLE"
+#define APP_SCHEDULER_TABLE_NAME        "SCHEDULER_TABLE"
+#define APP_DSCP_TO_TC_MAP_TABLE_NAME   "DSCP_TO_TC_MAP_TABLE"
+#define APP_QUEUE_TABLE_NAME            "QUEUE_TABLE"
+#define APP_PORT_QOS_MAP_TABLE_NAME     "PORT_QOS_MAP_TABLE"
+#define APP_WRED_PROFILE_TABLE_NAME     "WRED_PROFILE_TABLE"
```
Comment (eladraz) [inline common/scheme.h]: Should be in different commit.

File: common/scheme.h [hunk 7]
```diff
@@ -1,17 +1,32 @@
...
+
+
+
```
Comment (eladraz) [inline common/scheme.h]: Why so many empty lines?

File: common/select.cpp [hunk 8]
```diff
@@ -8,17 +10,28 @@ using namespace std;
...
+        return;
+    }
+    m_objects.push_back(selectable);
```
Comment (eladraz) [inline common/select.cpp]: I think it should be ASSERT() and in DEBUG mode only. There is no point for runtime check like that in production.

File: common/select.cpp [hunk 9]
```diff
@@ -8,17 +10,28 @@ using namespace std;
...
+}
+    
+void Select::addSelectables(_in_ std::vector<Selectable *> &selectables)
+{
+    for(auto it : selectables) {
+        addSelectable(it);
+    }
```
Comment (eladraz) [inline common/select.cpp]: I'm against these type of changes. I think that the Select class should do one thing, and adding vector of Selectable * sin't one of them.

File: common/select.cpp [hunk 10]
```diff
@@ -8,17 +10,28 @@ using namespace std;
...
+    }    
 }
 
 void Select::addFd(int fd)
 {
     m_fds.push_back(fd);
 }
 
-int Select::select(Selectable **c, int *fd, unsigned int timeout)
+int Select::select(_out_ Selectable **c, _out_ int *fd, _in_ unsigned int timeout)
```
Comment (eladraz) [inline common/select.cpp]: @stcheng are you ok with the _out_ _in_ weird convention?

File: common/select.h [hunk 11]
```diff
@@ -7,13 +7,15 @@
 #include <hiredis/hiredis.h>
 #include "selectable.h"
 
+
 namespace swss {
```
Comment (eladraz) [inline common/select.h]: Why the extra space?

File: common/table.h [hunk 12]
```diff
@@ -38,11 +38,11 @@ class Table {
...
 
 protected:
     /* Return the actual key name as a comibation of tableName:key */
     std::string getKeyName(std::string key);
-
```
Comment (eladraz) [inline common/table.h]: This should be a new patch

File: common/select.cpp (hunk 10 above)
Comment (stcheng) [inline common/select.cpp]: for my point of view, i don't think the #define in #define out #define inout is necessary in the code. the type itself indicates that.

File: common/consumertable.cpp (hunk 5 above)
Comment (hrachya-mellanox) [inline common/consumertable.cpp]: It really depends only OS as well. In past I had issues with this. I checked, it seems for Linux its ok.

File: common/select.cpp (hunk 10 above)
Comment (hrachya-mellanox) [inline common/select.cpp]: You can probably assume that Selectable **c should be out parameter, but int *fd wouldn't tell at all.
I don't think it is wierd, but usefull making declarations obviously clear, and self documenting.
Usually OS APIs, and other low level code have such annotations, especially when there is no documentation.
SAI has this as well.
In Windows I believe there are static analysis tools which will treat absence of in/out as error. I understand this is not Windows, but it has benefits outlined above.

**Comment by stcheng** [comment]
thanks
